arroyo-worker/src/connectors/source_collector.rs

```diff
@@ -0,0 +1,46 @@
+use crate::engine::Context;
+use crate::metrics::TaskCounters;
+use crate::rate_limiter::RateLimiter;
+use arroyo_formats::SchemaData;
+use arroyo_rpc::formats::BadData;
+use arroyo_types::{Key, Record, SourceError, UserError};
+use std::time::SystemTime;
+use tracing::warn;
+
+/// Collects a source record, handling errors and rate limiting.
+/// Considers the `bad_data` option to determine whether to drop or fail on bad data.
+pub async fn collect_source_record<K: Key, T: SchemaData>(
```

Comment:
Similarly here — this doesn't need to be its own module for one function

Reply:
Maybe this is actually better as a method on Context? Since we're taking in a &mut Context as the first argument anyways

Reply:
Good idea